### PR TITLE
[WI-46679] Add ext-parallel stubs

### PR DESCRIPTION
**Pull request by WyriHaximus**
Been using [`ext-parallel`](https://github.com/krakjoe/parallel/) for a while now and the missing stubs are annoying. This PR resolves: https://youtrack.jetbrains.com/issue/WI-46679

### CODE REVIEW

**Comment by kelunik** [comment]
How about making that a composer package? Works quite well for ext-async.

**Comment by WyriHaximus** [comment]
> How about making that a composer package? Works quite well for ext-async.

Could do that, but you can also composer require this repository [`jetbrains/phpstorm-stubs`](https://github.com/JetBrains/phpstorm-stubs/blob/master/composer.json) at `dev-master` once it's merged. But if you don't like/prefer that I'll set it up after I finished the next ext stub :smile: .

File: parallel/parallel/Events/Event/Type.php [hunk 1]
```diff
@@ -2,23 +2,27 @@
...
-    public const Cancel = 4;
+    public const Cancel = null;
 
     /* Runtime executing Event::$object (Future) was killed */
-    public const Kill = 5;
+    public const Kill = null;
 
     /* Event::$object (Future) raised error */
-    public const Error = 6;
+    public const Error = null;
```
Comment (dktapps) [inline parallel/parallel/Events/Event/Type.php]: This might cause problems with typehints and type inference.

Comment (WyriHaximus) [reply parallel/parallel/Events/Event/Type.php]: Good point since they are assumed to be an int, will set them to `0` instead.

Comment (WyriHaximus) [reply parallel/parallel/Events/Event/Type.php]: @dktapps Ok changed into a docblock instead, do you think that would solve the possible problems with type hints and type inference?

Comment (WyriHaximus) [reply parallel/parallel/Events/Event/Type.php]: FYI related issue https://github.com/krakjoe/parallel/issues/37#issuecomment-493502459

Comment (dktapps) [reply parallel/parallel/Events/Event/Type.php]: I don't know. I'd have to test.

File: parallel/parallel/Future/Error/KIlled.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?php
```
Comment (GrahamCampbell) [inline parallel/parallel/Future/Error/KIlled.php]: Typo in the file name here.

Comment (WyriHaximus) [reply parallel/parallel/Future/Error/KIlled.php]: Good catch, thanks fixed it :+1:

File: parallel/parallel/Events/Event/Type.php (hunk 1 above)
Comment (WyriHaximus) [reply parallel/parallel/Events/Event/Type.php]: Tested it just now and PHPStorm is fine with it this way.

**Comment by WyriHaximus** [comment]
FYI there is some ongoing work on `ext-parallel` fixing some small things now that `1.0.0` is out. Will update this PR later this week when the pace settles down there or do a follow up if this gets merged before that.

**Comment by MaXal** [comment]
@WyriHaximus Any news regarding updating the stubs after fixes in `ext-parallel`? They are on version `v1.1.2` right now.

**Comment by dktapps** [comment]
Quick Q: are single-file stubs preferred, or split ones like this?

**Comment by WyriHaximus** [comment]
@MaXal started working on this again yesterday evening, should have it done tonight

**Comment by MaXal** [comment]
@dktapps It doesn't matter much. Single file stubs can be a bit faster since they require less hard drive accesses.

**Comment by WyriHaximus** [comment]
@dktapps @MaXal as consumer I prefer one class per because it's easier to browse around in them. For functions I don't care much tbh.

**Comment by WyriHaximus** [comment]
@MaXal ok done. Added the missing Sync from `1.1`, the rest of the changes are bug fixes. Also ensured the code is valid PHP 7.0 code :smile: .

**Comment by Frago9876543210** [comment]
Current stubs incompatible with parallel
1. Why it should be "is valid PHP 7.0 code" when parallel supports 7.2+?
2. What about anonymous channels?
3. `Events\Event\Type` should not be null because `Event->type` will be just magic number. What can confuse you when you do var_dump of event

**Pull request merged**